ext/http/http_next.rs

```diff
@@ -0,0 +1,800 @@
+    AsyncRefCell::new(None),
+    CancelHandle::new(),
+    AsyncRefCell::new(rx),
+  ));
+  let cancel_clone = resource.cancel_handle();
+
+  let handle = spawn_local(SafeFutureForSingleThread(Box::pin(async move {
+    loop {
+      let conn = listener
+        .accept()
+        .try_or_cancel(cancel_clone.clone())
+        .await?;
+      serve_http_on(conn, &listen_properties, cancel_clone.clone(), tx.clone());
+    }
+    #[allow(unreachable_code)]
+    Ok::<_, AnyError>(())
+  })));
+
+  // Set the handle after we start the future
+  *RcRef::map(&resource, |this| &this.0)
+    .try_borrow_mut()
+    .unwrap() = Some(handle);
+
+  Ok(state.borrow_mut().resource_table.add_rc(resource))
+}
+
+#[op(v8)]
+pub fn op_serve_http_on<'scope>(
+  scope: &mut v8::HandleScope<'scope>,
+  state: Rc<RefCell<OpState>>,
+  conn: ResourceId,
+  init_cb: Value<'scope>,
+) -> Result<ResourceId, AnyError> {
+  let network_stream =
+    take_network_stream_resource(&mut state.borrow_mut().resource_table, conn)?;
+
+  let local_address = network_stream.local_address()?;
+  let listen_properties = DefaultHttpRequestProperties::listen_properties(
+    network_stream.stream(),
+    &local_address,
+  );
+  let init_callback: v8::Local<v8::Function> =
+    v8::Local::<v8::Function>::try_from(init_cb.v8_value)?;
+  let recv = v8::undefined(scope);
+  let scheme = v8::String::new(scope, listen_properties.scheme).unwrap();
+  let fallback_host =
+    v8::String::new(scope, listen_properties.fallback_host.as_str()).unwrap();
+  init_callback.call(
+    scope,
+    recv.into(),
+    &[scheme.into(), fallback_host.into()],
+  );
+
+  let (tx, rx) = tokio::sync::mpsc::channel(10);
+  let resource: Rc<HttpJoinHandle> = Rc::new(HttpJoinHandle(
+    AsyncRefCell::new(None),
+    CancelHandle::new(),
+    AsyncRefCell::new(rx),
+  ));
+
+  let handle = serve_http_on(
+    network_stream,
+    &listen_properties,
+    resource.cancel_handle(),
+    tx,
+  );
+
+  // Set the handle after we start the future
+  *RcRef::map(&resource, |this| &this.0)
+    .try_borrow_mut()
+    .unwrap() = Some(handle);
+
+  Ok(state.borrow_mut().resource_table.add_rc(resource))
+}
+
+#[op]
+pub async fn op_http_wait(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+) -> Result<u32, AnyError> {
+  // We will get the join handle initially, as we might be consuming requests still
+  let join_handle = state
+    .borrow_mut()
+    .resource_table
+    .get::<HttpJoinHandle>(rid)?;
+
+  let cancel = join_handle.clone().cancel_handle();
+  let next = async {
+    let mut recv = RcRef::map(&join_handle, |this| &this.2).borrow_mut().await;
+    recv.recv().await
+  }
+  .or_cancel(cancel)
+  .unwrap_or_else(|_| None)
+  .await;
+
+  //println!("got lock");
+  match next {
+    None => {
+      //println!("end");
+      let res = RcRef::map(join_handle, |this| &this.0)
+        .borrow_mut()
+        .await
+        .take()
+        .unwrap()
+        .await?;
+
```

Comment:
Nitpick: use `if Some()` here and return early? Makes it harder to read with all the indents